### DISCUSSION TITLE
Can't export my custom theme - VS Code

### DISCUSSION THREAD

**Discussion by jamosport** [marp-team/marp#288] [Q&A] [answered]
Hello, 
I've customed a css theme for marp in VS Code. In the interface, it works perfectely, but when i export my presentation, the layout disappears (in HTML and PDF). 

```
---
marp : true
theme: custom-theme
---
# Test
```
![image](https://user-images.githubusercontent.com/8986593/162063156-f2d34999-014b-4106-9ebf-3ee5fbcafe15.png)
![image](https://user-images.githubusercontent.com/8986593/162064107-13de9bf7-c484-4e96-92b4-92f75f4761f2.png)
```
/* custom-theme.css */
/* @theme custom-theme */

@import 'default';

section {
  background: grey;
}
```

![image](https://user-images.githubusercontent.com/8986593/162063354-d48cd96f-69cd-4e8b-8796-5356f86f95b0.png)

But, only the default theme is loaded, right ? 

Thanks in advance
Max

**Comment by yhatt** [comment]
Have located your Markdown file in the same workspace as the custom theme? The actual workspace that the Markdown document belongs to is often different from the current workspace that is opened by VS Code. (e.g. Opening the Markdown file located out of the current workspace)

To detect whether the custom theme was recognized by the Markdown document, check that there is no error in the definition of the `theme` directive.

![Untitled-2 md - marp - Visual Studio Code 2022_04_08 0_36_53](https://user-images.githubusercontent.com/3993388/162238401-fe86f5a5-846d-4cd2-b492-63fd5e545357.png)

In this case, I'm in the `~/Documents/marp` as the current workspace and opening 2 Markdown files located in `~/Documents/marp` and `~/Documents`. The former uses the custom theme from `~/Documents/marp/custom-theme.css` and the other uses `~/Documents/custom-theme.css` that was not found.

**Reply by jamosport** [reply]
Yes, the Md file is located in the same workspace as the custom css. Then, there is no error in the definition of the theme directive : 

![image](https://user-images.githubusercontent.com/8986593/162284180-7ae2f8f6-76b5-4530-ba4e-e4806c2703c2.png)

But, in the HTML file there is no link to the custom css. All the layout is in head section : 
![image](https://user-images.githubusercontent.com/8986593/162284443-79a69944-79c7-4a09-ad3e-3c32f842169b.png)

Thanks !

**Reply by yhatt** [reply]
It's a correct behavior that there is no external links (`<link href="">`) to custom theme CSS in the output. For portability of the export output, Marp compiles Markdown styles including custom theme CSS and inline styles into a one style used by `<style>` in `<head>`.

If looked like recognized the custom theme CSS by Marp correctly, try to check whether your CSS is valid by using [W3C CSS validator](https://jigsaw.w3.org/css-validator). Invalid styles included in both of custom theme and Markdown inline styles may effect to the Marp output.

> For debug the output CSS, it's good to export HTML through Marp CLI with configuration to disable minify for CSS (Refer to [Marp Core constructor options](https://github.com/marp-team/marp-core#constructor-options)).
https://github.com/marp-team/marp-cli#example
> 
> ```jsonc
> {
>   "options": {
>     "minifyCSS": false
>   }
> }
> ```
> 
> ```bash
> npx @marp-team/marp-cli@latest -c ./marp.config.json --theme ./custom-theme.css ./your-markdown.md
> ```

**Reply by jamosport** [reply]
According to the CSS Validation Service, my file is valid.

On Windows, it works with this command  :
`marp ./marp.config.json --theme ./custom-theme.css ./your-markdown.md`

The HTML file contains HTML tags now (that i need it). According to [Marp Core constructor options](https://github.com/marp-team/marp-core#constructor-options), it seems normal, right ? May i bypass this ?
![image](https://user-images.githubusercontent.com/8986593/162439388-7293ee5d-917f-489e-879c-449a5f54cfac.png)
![image](https://user-images.githubusercontent.com/8986593/162439465-ffb37530-bb1c-4df8-9c72-3c0d87f000c3.png)

It's almost resolved ! Thanks

**Reply by yhatt** [reply]
In Marp CLI, adding `--html` option to turn on raw HTML tags in Markdown would help your case.

BTW we still cannot reproduce your reported problem. If you want to resolve problem of export in VS Code for your better workflow, please try to re-install Marp extension and share logs outputted to the developer console while exporting. ([Help] -> [Toggle Developer Tools] -> [Console] tab)

**Comment by jamosport** [comment, marked as answer]
I'd like to thank you for your help. This command works for me : `marp ./marp.config.json --theme ./custom-theme.css ./myfile.md --html`

**Reply by yhatt** [reply]
OK, marked as answer provisionally.

Using Marp CLI is just a workaround, and we are still welcome follow up for the original problem about not loaded custom theme in the export output of VS Code extension.

**Comment by jamosport** [comment]
Hello,
This time, I tried to export my slides to PDF :
`marp ./marp.config.json --theme ./custom-theme.css ./myfile.md --pdf`

But it doesn't work. As one of my last post, the PDF file contains HTML tags too. I tried other options, but i obtain the same result.

Thanks in advance,

**Reply by yhatt** [reply]
The provided command seems to be missing `--html` option again. Try to add `--html` for rendering raw HTML tags.

`--pdf` and `--html` can set at the same time, because `--html` option does not mean to specify the export file type.

**Reply by jamosport** [reply]
OK, thanks. It works ! I hadn't thought about it...

**Comment by mahomedalid** [comment]
I am having the exact same issue. When executing `marp -s .` theme is loaded. If I export as `marp mypresentation.md -o output.html` it does not work. If I pass the parameters for the config and explicitly `--theme mytheme.css`  it works. My theme is specified in the header of the markdown.

I am using marp cli 4.2.3 I will try to upload my presentation to a public repo if that helps.

**Reply by yhatt** [reply]
`marp -s .` is equivalent to `marp -s --input-dir .`. [`--input-dir`](https://github.com/marp-team/marp-cli#:~:text=string-,%2D%2Dinput%2Ddir,-%2DI) automatically loads theme CSS files in the specified folder.

If tried to convert a single Markdown file, the custom theme file will not load because you do not specify loading theme CSS files. To load the custom CSS in the current directory, you have to specify [`--theme-set` option](https://github.com/marp-team/marp-cli#theme-set) as `.`.

```bash
marp --theme-set . -- mypresentation.md -o output.html
```

**Reply by mahomedalid** [reply]
Thanks! That makes sense @yhatt.

I still tried with

```
marp .
```

but it did not work. I had to run

```
marp . --theme-set .
```

to get the CSS loaded. So I believe there is still some inconsistency or at least a confusing UX here. The same directory behaves differently depending on whether you use `-s` or not, and without an explicit flag it is not obvious why the theme is ignored.

[This is my theme/presentation.](https://github.com/mahomedalid/presentations/blob/main/badgefed/fosdem2026/README.md)

**Reply by yhatt** [reply]
The regular conversion mode `marp .` and server mode `marp -s .` take different kinds of input `.`.

In the regular conversion mode, **`.` means `./**/*.md`, the multiple Markdown files in the current directory.** _This path is not for specifying the location of CSS themes._ Thus, to use custom theme CSS files, you have to use with an additional `--theme-set` option as `.` (it means `./**/*.css`)

On the other hand, [the server mode `--server`/`-s` must specify **a path for the single directory**](https://github.com/marp-team/marp-cli#:~:text=We%20require%20to%20pass%20%2D%2Dserver%20(%2Ds)%20option%20and%20a%20directory%20to%20serve.), to serve contents on the local server `http://localhost:8080`. In this case, both of Markdown and theme CSS will be found from the specified directory so you do not need `--theme-set` option. (This rule also applied in the regular conversion mode with [`--input-dir`/`-I` option](https://github.com/marp-team/marp-cli#:~:text=When%20you%20want%20to%20output%20the%20converted%20result%20to%20another%20directory%20with%20keeping%20the%20origin%20directory%20structure%2C%20you%20can%20use%20%2D%2Dinput%2Ddir%20(%2DI)%20option.))
